### ISSUE TITLE
Cannot connect to the GPM

### ISSUE THREAD

**Issue by FrenchW** [getgrav/grav#850]
Hi, 
On each admin page, I have the red popup : "Cannot connect to the GPM"
I've red the help page here https://learn.getgrav.org/troubleshooting/common-problems and if it does not resolve the issue, it's said to keep in touch with you.

Hosting is in an OVH.com shared server (said "Mutualisé" in french)

# Configuration
## System
`System : Linux webm333.start.ha.ovh.net 3.14.60-grsec-hosting-web-3.14 #1 SMP Wed Feb 17 13:41:21 UTC 2016 x86_64`
`Build Date : Apr 29 2016 07:00:47`
`Configure Command : './configure' '--with-openssl' '--with-mysqli=mysqlnd' '--with-pdo-mysql=mysqlnd' '--prefix=/usr/local/php7.0' '--enable-fpm' '--enable-opcache' '--with-gd' '--enable-intl' '--enable-gd-native-ttf=/usr' '--with-png-dir=/usr' '--enable-wddx' '--with-config-file-path=/usr/local/php7.0/etc' '--enable-sigchild' '--enable-short-tags' '--disable-rpath' '--enable-libgcc' '--with-zlib=/usr' '--enable-bcmath' '--enable-calendar' '--with-curl=/usr' '--with-gdbm=/usr' '--enable-ftp' '--with-gettext=/usr' '--with-imap' '--with-kerberos' '--with-imap-ssl' '--with-mcrypt=/usr' '--with-mhash=/usr' '--with-jpeg-dir=/usr' '--enable-exif' '--enable-sysvsem' '--enable-sysvshm' '--enable-sysvmsg' '--enable-zip' '--enable-inline-optimization' '--enable-soap' '--enable-mbstring' '--enable-mbregex' '--with-gnu-ld' '--with-libxml-dir' '--enable-shared=yes' '--enable-static=yes' '--with-xsl=/usr' '--enable-sockets' '--with-xmlrpc' '--with-iconv' '--with-imagick' '--enable-pdo' '--with-pdo-pgsql=/usr' '--with-xpm-dir=/usr' '--with-pcre-regex' '--with-freetype-dir=/usr' '--with-pspell' '--with-pgsql=/usr' '--with-bz2' '--with-gmp' 'EXTENSION_DIR=/usr/local/php7.0/lib/php-extensions'`

## Curl
`cURL support : enabled`
`cURL Information : 7.38.0`
`Age : 3`

## OpenSSL
`OpenSSL support : enabled`
`OpenSSL Library Version : OpenSSL 1.0.1k 8 Jan 2015`
`OpenSSL Header Version : OpenSSL 1.0.1k 8 Jan 2015`

## Proxy
No proxy with OVH

## Permissions 
No problem, All writable and in bin, all files also executables (755).

# Conclusion
Except the GPM alert, all works fine. 

# May be related ?
Just an error when trying to

- Add a theme (clicked on add):**` Call to a member function filter() on null`**

> return $local ? $gpm->getInstalledPlugins() : $gpm->getRepositoryPlugins()->filter(function (
> **…/­user/­plugins/­admin/­classes/­admin.php 494**

- Add a plugin: **`Call to a member function filter() on null`**

> return $local ? $gpm->getInstalledThemes() : $gpm->getRepositoryThemes()->filter(function ($package, $slug) use
> …/­user/­plugins/­admin/­classes/­admin.php 471

Thanks in advance
Fred

**Comment by rhukster** [comment]
Does GPM work from the console?  Ie, from Grav root:

```
$ bin/gpm index
```

**Comment by FrenchW** [comment]
When I'm using it right from my mac, your shell command works.
When I'm using if with Mamp locally, the website shows no error. But, when I'm transferring the whole file to the distant web server, it starts showing "Cannot connect to the GPM" as discussed previously"

**Comment by flaviocopes** [comment]
Yep, the problem is on the server. Do you have a SSH available on that server?

**Comment by FrenchW** [comment]
no SSH on the server. 
This could be told by a switch somewhere in the configuration to tell Grav not to try it.
Because the alert appears on each page refresh, it's a bit disturbing when you don't know how the app works. And also, if ssh is a prequisite, it must be told in the documentation and the troubleshooting part that tries to resolve this issue

**Comment by rhukster** [comment]
BTW, do you have a dedicated account? If so you should have SSH access: http://help.ovh.com/SshOnDedicated

**Comment by FrenchW** [comment]
Hi, thanks for your answer. 
Setting Off "Check for updates" solves the message appearance.
Thank you ! This should be added to your troubleshooting doc (I mean: "if no ssh .... blah blah) :)

Are those errors also relative to ssh ? : Just an error when trying to

Add a theme (clicked on add):Call to a member function filter() on null
return $local ? $gpm->getInstalledPlugins() : $gpm->getRepositoryPlugins()->filter(function (
…/­user/­plugins/­admin/­classes/­admin.php 494
Add a plugin: Call to a member function filter() on null
return $local ? $gpm->getInstalledThemes() : $gpm->getRepositoryThemes()->filter(function ($package, $slug) use
…/­user/­plugins/­admin/­classes/­admin.php 471

**Comment by flaviocopes** [comment]
The problem is Grav cannot connect to the GPM (Grav Package Manager). By turning off the "check for updates", you don't see the error pop up, but themes/plugins cannot be installed or updated, nor Grav itself. Could be a limitation imposed on your web server by the hosting provider, so cURL towards an external service fails.

**Comment by FrenchW** [comment]
I understand, no problem. But i'm not a symphony, nor php developer. But as I can see, a CMS like wordpress is able to update itself (meant : connect to other site, download file, extract them etc etc)
Even on OVH, il works like a charm without ssh. 
Those 2 errors could be avoided,a least with a try/catch. I suppose ...

**Comment by rhukster** [comment]
We have seen this issue, and we know it exists, but we just need to track down what exactly is causing it.    Is this account of yours production? can we perhaps get access to it and try to debug this in a test subdirectory?

**Comment by thwillingham** [comment]
I have this same issue.  I installed Grav without the admin plugin.  When I tried to install the admin plugin via the terminal, I got an error.  Long story short, I have to use '$ php-cli bin/gpm'  to run commands.  I can only assume that my 'cannot connect to gpm' error is related.

**Comment by flaviocopes** [comment]
@thwillingham did you check https://learn.getgrav.org/troubleshooting/common-problems#cannot-connect-to-the-gpm  ?

**Comment by thwillingham** [comment]
Thanks @flaviocopes!  It ended up being a permissions issue.  The script on https://learn.getgrav.org/troubleshooting/permissions fixed the problem.

**Comment by bennigraf** [comment]
Hey,
for future reference: On my local dev machine I had the same issue with the openssl php extension being installed but still not being able to access gpm. The reason was there was no cert file configured to compare the gpm certificate against. I had to add something like 
`openssl.cafile=/opt/local/share/curl/curl-ca-bundle.crt`
to my php.ini to make it work (I installed openssl & curl via MacPorts).
